lib/auth/join_tpm.go

```diff
@@ -0,0 +1,118 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"context"
+	"crypto/x509"
+	"log/slog"
+
+	"github.com/google/go-attestation/attest"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/modules"
+	"github.com/gravitational/teleport/lib/tpm"
+)
+
+func (a *Server) RegisterUsingTPMMethod(
+	ctx context.Context,
+	initReq *proto.RegisterUsingTPMMethodInitialRequest,
+	solveChallenge client.RegisterTPMChallengeResponseFunc,
+) (*proto.Certs, error) {
+	// First, check the specified token exists, and is a TPM-type join token.
+	if err := initReq.JoinRequest.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	pt, err := a.checkTokenJoinRequestCommon(ctx, initReq.JoinRequest)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ptv2 := pt.(*types.ProvisionTokenV2)
```

Comment:
I really dislike the massive interface we already have over ProvisionTokenV2, I think with all the join methods, you'd end up with tonnes and tonnes of getter methods, it'd also be super weird on the interface because they'd all need to return unimplemented for ProvisionTokenV1. I'll add a cast err catch.

Reply:
Any anti-interface argument is super effective with me. I'm happy with the `ok` check.

Reply:
e65fe163b3